commons-math/src/main/java/org/apache/commons/math4/special/BesselJ.java

```diff
@@ -0,0 +1,649 @@
+     * that can be considered accurate.
+     * </p><p>
+     * <ul>
+     * <li>nVals < 0: An argument is out of range. For example, nb <= 0, alpha
+     * < 0 or > 1, or x is too large. In this case, b(0) is set to zero, the
+     * remainder of the b-vector is not calculated, and nVals is set to
+     * MIN(nb,0) - 1 so that nVals != nb.</li>
+     * <li>nb > nVals > 0: Not all requested function values could be calculated
+     * accurately. This usually occurs because nb is much larger than abs(x). In
+     * this case, b(n) is calculated to the desired accuracy for n < nVals, but
+     * precision is lost for nVals < n <= nb. If b(n) does not vanish for n >
+     * nVals (because it is too small to be represented), and b(n)/b(nVals) =
+     * \(10^{-k}\), then only the first NSIG-k significant figures of b(n) can be
+     * trusted.</li></ul></p>
+     */
+    public static class BesselJResult {
+
+        /** Bessel function values */
+        private final double[] vals;
+
+        /** Valid value count */
+        private final int nVals;
+
+        /**
+         * Create a new BesselJResult with the given values and valid value count.
+         *
+         * @param b values
+         * @param n count of valid values
+         */
+        public BesselJResult(double[] b, int n) {
+            vals = MathArrays.copyOf(b, b.length);
+            nVals = n;
+        }
+
+        /**
+         * @return the computed function values
+         */
+        public double[] getVals() {
+            return MathArrays.copyOf(vals, vals.length);
+        }
+
+        /**
+         * @return the number of valid function values (normally the same as the
+         *         length of the array returned by {@link #getnVals()})
+         */
+        public int getnVals() {
+            return nVals;
+        }
+    }
+
+    /**
+     * Calculates Bessel functions \(J_{n+alpha}(x)\) for
+     * non-negative argument x, and non-negative order n + alpha.
+     * <p>
+     * Before using the output vector, the user should check that
+     * nVals = nb, i.e., all orders have been calculated to the desired accuracy.
+     * See BesselResult class javadoc for details on return values.
+     * </p>
+     * @param x non-negative real argument for which J's are to be calculated
+     * @param alpha fractional part of order for which J's or exponentially
+     * scaled J's (\(J\cdot e^{x}\)) are to be calculated. 0 <= alpha < 1.0.
+     * @param nb integer number of functions to be calculated, nb > 0. The first
+     * function calculated is of order alpha, and the last is of order
+     * nb - 1 + alpha.
+     * @return BesselJResult a vector of the functions
+     * \(J_{alpha}(x)\) through \(J_{nb-1+alpha}(x)\), or the corresponding exponentially
+     * scaled functions and an integer output variable indicating possible errors
+     */
+    public static BesselJResult rjBesl(double x, double alpha, int nb) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 61. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.